### DISCUSSION TITLE
CRUD e Entity Framework - ViewModels

### DISCUSSION THREAD

**Discussion by DOSREI** [balta-io/forum#655] [Tira Dúvidas] [answered]
Gostaria de entender qual a diferença de ViewModel para DTO.

**Comment by andrebaltieri** [comment, marked as answer]
Bom dia, @DOSREI, como vai?

Legal demais, é uma ótima questão e que realmente deixa muitas dúvidas.
Gostaria só de ouvir antes, na sua opinão, quais seriam as diferenças entre elas, o que você acha?
Não precisa estar certo ou errado, é só um exercício mesmo!

**Reply by DOSREI** [reply]
Boa noite desculpas a demora.
Andei dando uma pesquisa vi basicamente um trata o que vem da tela outro do banco. 
Mas sinceramente não encontrei nenhuma resposta que deixasse totalmente claro.

**Reply by andrebaltieri** [reply]
Perfeito, já é um começo... e nas suas pesquisas, qual trata o que vem da tela e qual trata o que vem do banco?
Quais canais e fontes você pesquisou?

**Reply by DOSREI** [reply]
Dto trata dos dados do banco
Views trata os dados da tela. 

Foi em varias fontes, mas a melhor explicação (com mais clareza), estava no stackoverflow.

**Reply by andrebaltieri** [reply]
Bom dia, @DOSREI, como vai?
Não necessariamente.

DTO é a sigla para Data Transfer Object (Objeto de transferência de dados), o que significa que ele é apenas uma forma de transportar dados de um lugar para outro. Pode ser do banco para aplicação, entre as camadas da aplicação e até externalisar os dados.

ViewModel, como o nome sugere é a junção de dois itens, a View e o Model. A View é a toda parte visível ou que será usada na parte visível e o Model normalmente se refere a dados (Entidades e outros modelos).

A forma mais simples de te explicar um ViewModel é usando o exemplo de cadastro de usuário... Para cadastrar um usuário, precisamos do usuário e senha... o banco só precisa destas informações.

Porém, na tela, normalmente temos a confirmação de senha, um campo usado apenas na View, para aferição das senhas digitadas.
Neste caso, teríamos o seguinte:
```csharp
// Este seria o modelo
public class User 
{ 
    public string UserName { get; set; }
    public string Password { get; set; }
} 

// Esta seria a ViewModel
public class User 
{ 
    public string UserName { get; set; }
    public string Password { get; set; }
    public string ConfirmPassword { get; set; }
} 
```

Em resumo a ViewModel é uma classe que fica entre a tela e o modelo, para desacoplar a tela de um modelo específico, já que nem sempre os dados exibidos na tela são necessariamente os dados presentes em um model.

Agora te pergunto... se vamos trafegar a ViewModel entre camadas, até chegar lá próximo do banco, isto também não torna ela uma DTO? 😆
